Extending pipeline promotions token security documentation (#3540)
weaveworks/weave-gitops: website/docs/pipelines/promoting-applications.mdx

> Permissions are purely additive (there are no "deny" rules).

You should ensure that attempts to break this contract are blocked and detected. You could achieve it
by using Weave GitOps' [Policy capabilities](../../policy/intro/). Policy Agent acts in two complementary modes:
- [Admission Controller](../../policy/intro#admission-controller) protects from any attempt to create non-compliant RBAC resources that would end granting access to the secret.
- [Audit](../../policy/intro#audit) helps you identify already existing resources that are out of compliance. For example,
roles created before policy agent was introduced as admission controller.

Once you have enabled Policy, the [Policy Library](../../policy/weave-policy-profile/) gives you a set of
good practices policies that will help you securing your pipeline secrets according to previous RBAC recommendations. Deploy
them as Kustomization based on the following example:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1
kind: GitRepository
metadata:
  name: policy-library
  namespace: flux-system
spec:
  interval: 10m0s
  url: https://github.com/weaveworks/policy-library.git

>   secretRef:

Author: review with @MostafaMegahid whether this is the best way to use it or we should use other sources

Author: To review today with Mostafa.

Author: From the conversation with Mostafa: 

- We expect policy library to be used as a set of examples that customers might need to customize.
- The way we expect now to do so is via Git, copy and paste then to update with the values that better suit them.
- Library is enterprise and its value would be protected by the ww - customer agreement.

Reviewer: How do people get access to it?

Author: To follow up next week in policy distribution out of git

is upgraded in one of the pipeline's environments, a PR is created that updates the manifest file setting the application version in the next environment.

The dynamic nature of GitOps deployments requires you to assist Weave GitOps a little with information on which repository hosts the manifest files,
how to authenticate with the repository and the Git provider API, and which file hosts the version definition for each environment.

### Security

#### Environments and Repositories

1. Use it in low-risk environments and not in production: pipelines is an alpha feature and not yet production-ready.
2. Make sure you have considered possible attack vectors to production, and put controls in place. Some
of these scenarios are:
- In the case of a [monorepo](https://fluxcd.io/flux/guides/repository-structure/#monorepo): you want to ensure that a compromised token doing a change in test cannot end up doing a promotion in production without a pull request review, for example, leveraging [Codeowners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners).

> - In the case of [repo per environment](https://fluxcd.io/flux/guides/repository-structure/#repo-per-environment) you want to ensure that a pipeline is configured with your test environment repo URL. You also want to ensure that you have seggregation of tokens per environment rather than allowing a token to access any environment repo.

Reviewer: ```suggestion
- In the case of [repo per environment](https://fluxcd.io/flux/guides/repository-structure/#repo-per-environment) you want to ensure that a pipeline is configured with your test environment repo URL. You also want to ensure that you have segregation of tokens per environment rather than allowing a token to access any environment repo.
```

#### Policy

By following the guidelines above, you can have a safe initial configuration. However, given there are no deny semantics in [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole), you need to guard future changes.

> An RBAC Role or ClusterRole contains rules that represent a set of permissions.
> Permissions are purely additive (there are no "deny" rules).

You should ensure that attempts to break this contract are blocked and detected. You could achieve it
by using Weave GitOps' [Policy capabilities](../../policy/intro/). Policy Agent acts in two complementary modes:
- [Admission Controller](../../policy/intro#admission-controller) protects from any attempt to create non-compliant RBAC resources that would end granting access to the secret.
- [Audit](../../policy/intro#audit) helps you identify already existing resources that are out of compliance. For example,
roles created before policy agent was introduced as admission controller.

Once you have enabled Policy, the [Policy Library](../../policy/weave-policy-profile/) gives you a set of

> good practices policies that will help you securing your pipeline secrets according to previous RBAC recommendations. Deploy

Reviewer: ```suggestion
good practices policies that will help you keep pipeline secrets secure according to the previous RBAC recommendations. Deploy
```

them as Kustomization based on the following example:

```yaml

> apiVersion: source.toolkit.fluxcd.io/v1

Author: provide guidance on how to leverage CREs for helping you

Author: updated https://github.com/weaveworks/weave-gitops/pull/3540/files#r1274777480

kind: GitRepository
metadata:
  name: policy-library
  namespace: flux-system
spec:
  interval: 10m0s
  url: https://github.com/weaveworks/policy-library.git
  secretRef:
    name: policy-library-github-credentials
---
apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  name: rbac-secrets-good-practices
  namespace: flux-system
spec:
  interval: 1m0s
  sourceRef:
    kind: GitRepository
    name: policy-library
  path: ./goodpractices/kubernetes/rbac/secrets
  prune: true
```

:::caution

Policies might have exclusion features to accommodate privileged workloads like Flux. You could manage them via [PolicyConfig](https://docs.gitops.weave.works/docs/policy/policy-configuration/).
For example, in order to allow Flux you could use the following `PolicyConfig`:

```yaml
apiVersion: pac.weave.works/v2beta2
kind: PolicyConfig
metadata:
  name: allow-flux
spec:
  match:

>     apps:

Author: This example shows how you could leverage PolicyConfig to add exclusions for privilege workloads for your policies. 

In our example, we are deploying the goodpractices kustomization that deploy policies like `weave.templates.rbac-prohibit-wildcards-policyrule-resources`. These policies by default does not have exclusions. In case you want to have exclusions (see my comment [here](https://github.com/weaveworks/policy-library/pull/31/files#r1273143158)) you leverage PolicyConfig.

      - kind: Kustomization
        name: flux-system
        namespace: flux-system
  config:
    weave.templates.rbac-prohibit-wildcards-policyrule-resources:
      parameters:
        exclude_label_key: "app.kubernetes.io/part-of"
        exclude_label_value: "flux"
    weave.templates.rbac-prohibit-wildcards-policyrule-verbs:
      parameters:
        exclude_label_key: "app.kubernetes.io/part-of"
        exclude_label_value: "flux"
    weave.policies.rbac-protect-list-secrets:
      parameters:
        exclude_label_key: "app.kubernetes.io/part-of"
        exclude_label_value: "flux"
    weave.policies.rbac-protect-watch-secrets:
      parameters:
        exclude_label_key: "app.kubernetes.io/part-of"
        exclude_label_value: "flux"
    weave.policies.rbac-protect-wildcard-secrets:
      parameters:
        exclude_label_key: "app.kubernetes.io/part-of"
        exclude_label_value: "flux"
```

> Remind not allowing users to create RBAC resources without compliance checks. Otherwise, they could create RBAC resources that could escape this runtime control.

Author: it tries to call out that if you have exclusions, they could be used as an escalation mechanism. Given exclusions appears in the doc first time here, we just try to call out / reinforce the risk and how to mitigate it.

> An RBAC Role or ClusterRole contains rules that represent a set of permissions.
> Permissions are purely additive (there are no "deny" rules).

You should ensure that attempts to break this contract are blocked and detected. You could achieve it
by using Weave GitOps' [Policy capabilities](../../policy/intro/). The Policy Agent acts in two complementary modes:
- [Admission Controller](../../policy/intro#admission-controller) protects from any attempt to create non-compliant RBAC resources that would end granting access to the secret.
- [Audit](../../policy/intro#audit) helps you identify already existing resources that are out of compliance. For example,
roles created before policy agent was introduced as admission controller.

Once you have enabled Policy, the [Policy Library](../../policy/weave-policy-profile/) gives you a set of
good practices policies that will help you keep pipeline secrets secure according to the previous RBAC recommendations. Deploy
them as Kustomization based on the following example:

:::tip
In case you don't have access to the Policy Library, work with your Weaveworks Technical Account Manager (TAM) or

> Weaveworks Customer Reliability Engineer (CRE) to help with this step.

Author: @squaremo let me know your thoughts

<img width="854" alt="Screenshot 2023-07-26 at 11 46 02" src="https://github.com/weaveworks/weave-gitops/assets/12957664/f109397d-4dfa-48e8-940a-2fba6e757e38">

Reviewer: If those are real job titles, I'm on board with it :-)

The dynamic nature of GitOps deployments requires you to assist Weave GitOps a little with information on which repository hosts the manifest files,
how to authenticate with the repository and the Git provider API, and which file hosts the version definition for each environment.

> ### Security

Author: Extract the general recommendations to a general-purpose page and reference it here. it is required for other cases.

Author: We will refactor after merged so we dont increase scope from the original one.

:::danger

Create pull requests requires write access to your git repo. In case your pull request secret is compromised (and it is
suggested you assume it as a likely scenario more than unlikely), it could end up for example, in deleting all your
production applications.

This section should not be read as best practices but as baseline to enable automated pull requests in lower environments.

It targets platform admins before enabling the capability.
:::

#### Environments

1. **Use it in a lower environment when there are no attack vectors to production**: avoid using it if a compromised token could
make changes in a lower environment (for example test or staging) that could be promoted to production being undetected
nor rejected.

#### RBAC

> You have configured RBAC following [Kubernetes](https://kubernetes.io/docs/concepts/security/rbac-good-practices/).

Author: Extend with ours too

>An RBAC Role or ClusterRole contains rules that represent a set of permissions.
> Permissions are purely additive (there are no "deny" rules).

You should ensure that attempts to break this contract are blocked and detected. You could achieve it
by leveraging [Policy capabilities](../../policy/intro/). Policy Agent acts in two complementary modes:
- [Admission Controller](../../policy/intro#admission-controller) protects from any attempt to create non-compliant RBAC resources that would end granting access to the secret.
- [Audit](../../policy/intro#audit) helps you identify already existing resources that are out of compliance. For example,
roles created before policy agent was introduced as admission controller.

Once enabled Policy, [Policy Library](../../policy/weave-policy-profile/) gives you have a set of policies to
help you with the task. `RBACProhibitWildcards` and `RBACProhibitVerbsOnResources` are the building blocks
that you could use to implement RBAC least privilege. Providing values to the policy parameters to restrict
wildcards and access to secrets:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:

> - policies/RBACProhibitWildcards/no-wildcard-on-resources-policy.yaml

Author: @MostafaMegahid how should we contribute these policies to the policy library ?

Author: Today we synced on this topic. the suggested path would be to 

1. add these policies to [policy library](https://github.com/weaveworks/policy-library/tree/main/policies). An example of this could https://github.com/weaveworks/policy-library/tree/main/policies/RBACClusterRoleClusterAdmin

Author: Draft PR for good practices within policy-library https://github.com/weaveworks/policy-library/pull/31

Author: To review today with Mostafa

The dynamic nature of GitOps deployments requires you to assist Weave GitOps a little with information on which repository hosts the manifest files,
how to authenticate with the repository and the Git provider API, and which file hosts the version definition for each environment.

### Security

:::danger

Create pull requests requires write access to your git repo. In case your pull request secret is compromised (and it is

> suggested you assume it as a likely scenario more than unlikely), it could end up for example, in deleting all your

Author: done

production applications.

> This section should not be read as best practices but as baseline to enable automated pull requests in lower environments.

Author: >lower environments 
low risk environment

Author: updated

It targets platform admins before enabling the capability.
:::

#### Environments

1. **Use it in a lower environment when there are no attack vectors to production**: avoid using it if a compromised token could

> make changes in a lower environment (for example test or staging) that could be promoted to production being undetected

Author: extend with questions that you should ask as platform admin to make it a workable scenario - for example shared configuration repo (test, staging, prod)

we cannot provide checklist but guidance

ex solution leverage codeowners for extra eyes for production  

to review in the context of the shared responsibility model

Author: updated

nor rejected.

#### RBAC

You have configured Least privilege RBAC following [Good Practices](https://kubernetes.io/docs/concepts/security/rbac-good-practices/).
Emphasis in the following:

> Assign permissions at the namespace level where possible. Use RoleBindings as opposed to ClusterRoleBindings
> to give users rights only within a specific namespace.

> Avoid providing wildcard permissions when possible, especially to all resources.
> As Kubernetes is an extensible system, providing wildcard access gives rights not just to
> all object types that currently exist in the cluster, but also to all object types which are created in the future.

> Avoid list and watch on secrets as:

Author: do not extract get of the equation

Author: what about the scenario where granting get access to a service account and someone init he namespace can run pods with that service account

Author: >what about the scenario where granting get access to a service account and someone init he namespace can run pods with that service account
added

> It is also important to note that list and watch access also effectively allow for users to reveal the Secret contents.

Test your RBAC configuration against privilege escalation using a scenario like the following one:

```gherkin
Cannot read pipeline secrets via RBAC privilege escalation

> Scenario Outline:

Author: provide running example

Author: what about templates

Author: adding https://github.com/weaveworks/weave-gitops-quickstart/pull/5

to implement the expected behaviour:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- policies/RBACProhibitWildcards/no-wildcard-on-resources-policy.yaml
- policies/RBACProhibitWildcards/no-wildcard-on-verbs-policy.yaml
- policies/RBACProhibitVerbsOnResources/no-list-secrets-policy.yaml
- policies/RBACProhibitVerbsOnResources/no-watch-secrets-policy.yaml
- policies/RBACProhibitVerbsOnResources/no-wildcard-secrets-policy.yaml
```

An example of policy to reject using wildcards (in this case using wildcards on verbs) is shown below:

```yaml
apiVersion: pac.weave.works/v2beta2
kind: Policy
metadata:
  name: weave.templates.rbac-prohibit-wildcards-verbs
spec:
  id: weave.templates.rbac-prohibit-wildcards-verbs
  name: Rbac Prohibit Wildcards on Verbs
  ...
  category: weave.categories.access-control
  severity: high
  parameters:
    - name: attributes
      type: string
      required: true
      value: "verbs"
    - name: exclude_role_name
      type: string
      required: true
      value: "wego-admin"
    - name: exclude_label_key
      type: string
      required: false
      value: "app.kubernetes.io/part-of"

>     - name: exclude_label_value

Author: extend 

1) be careful with rolebindings out of the namespace - for example, allowing, extending 
2) dont give kubectl access with rbac write access because as a user i could escape this control

Author: >dont give kubectl access with rbac write access because as a user i could escape this control

added

      type: string
      required: false
      value: "flux"
```

An example of policy to reject allowing list secrets is shown below:

```yaml
apiVersion: pac.weave.works/v2beta2
kind: Policy
metadata:
  name: weave.templates.rbac-prohibit-list-on-secrets
spec:
  id: weave.templates.rbac-prohibit-list-on-secrets
  name: Rbac Prohibit Wildcard On Secrets
  ...
  category: weave.categories.access-control
  severity: high
  parameters:
    - name: resource
      type: string
      required: true
      value: "secrets"
    - name: verb
      type: string
      required: true
      value: "list"
    - name: exclude_label_key
      type: string
      required: false
      value: "app.kubernetes.io/part-of"
    - name: exclude_label_value
      type: string
      required: false
      value: "flux"
```

An example of privilege escalation for pipelines blocked via Policy Agent is shown below

![privilege escalation blocked](img/pipeline-priviledge-escalation-blocked.png)

> #### Service Account

Author: controller should live in its namespace

Author: This concern it is reflected in the shared security context board. it will be part of the future security documentation.

```
</details>

At this stage, pipeline controller can read the promotion secrets, however given there are no deny semantics in [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole):

>An RBAC Role or ClusterRole contains rules that represent a set of permissions. Permissions are purely additive (there are no "deny" rules).

You would need to ensure that no other `rolebindings` or `clusterrolebndings` would allow read the promotion secret at any time.

This could be achieved in various ways. For example, by leveraging [policy capabilities](../../policy/intro/)
to detect existing and future creation of roles that would grant read secrets permissions.

<details><summary>Expand to see example</summary>

Once enabled [policy and policy library](../../policy/weave-policy-profile/) you have a set of available policies to
leverage. The one we are interested here is called `Rbac Prohibit Verbs On Resources` to restrict
the creation of roles that would allow access to the promotion credentials.

```yaml

> apiVersion: pac.weave.works/v2beta2

Author: ![Screenshot 2023-03-28 at 09 27 41](https://user-images.githubusercontent.com/12957664/228176447-b845b1a2-7afb-4cfb-bb5c-7bb346044864.png)

- policy or policy config https://docs.gitops.weave.works/docs/0.18.0/policy/policy-configuration/

kind: Policy
metadata:
    name: weave.templates.rbac-prohibit-verbs-on-resources
spec:
    id: weave.templates.rbac-prohibit-verbs-on-resources
    name: Rbac Prohibit Verbs On Resources
```
that we could provide values with
```yaml
parameters:
 - name: resource
   type: string
   required: true
   value: "secrets"
 - name: verb
   type: string
   required: true

>    value: "get" # do the same for list and watch

Reviewer 2: Can we do a couple of things here?
 1. Put the full version of this for users so that they can copy and paste.
 2. Check what happens when someone uses `*` for verbs

Author: Given policy library is enterprise feature, I cannot add it here.

and for security and auditing perspective, you don't want to impersonate a real user for it.

<details><summary>Expand to see example</summary>

![using bot account](img/bot-account.png)
</details>

2. **Do not use long-live tokens**: set an expiration date and rotate them according to your security policy.

<details><summary>Expand to see example</summary>

![create token with expiration](img/create-token-with-expiration.png)
</details>

3. **Honour the least privilege principle**: avoid having high privilege tokens. Restrict the token to your just your repo and to just the operations required.

<details><summary>Expand to see example</summary>

For example, if the case of GitHub, use [fine-grained tokens](https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/) to only
allow access to the single repo that your configuration manifests exist.

![create least privileged token](img/fine-grained-token.png)
 </details>

4. **Review the tokens on a regular basis following your git provider recommendations**. Ensure that:
    - Only the ones that are required are present.

>     - Tokens close to its expiration date are managed accordingly.

Reviewer: ```suggestion
    - Tokens close to their expiration date are cycled.
```

![create token least priviledge](img/fine-grained-token.png)
 </details>

> 4. **Review active access tokens on a regular basis**: to ensure that only the ones that are required are present at all times.

Author: extend with following guidance for git provider and need to update the secret after it expires.

Author: updated

Weave GitOps Enterprise will look for this marker whenever it receives an event from the respective HelmRelease of one of the leaf clusters and patch the file with the version denoted in the event (see the section above for instructions on setting up notification events from leaf clusters). Finally, it will create a Git provider PR to update the version of the application for the next environment in the pipeline.

> ### Supported Git Providers

Reviewer: I think from here down it's just moved text, is that right? (feel free to ignore comments below here!)

Author: Umms, I could not see comments for this section: feel free to add them to review them too. Thanks!

Reviewer: There was just one I think; I didn't bother reviewing most of the text that you didn't change.

The following Git providers are currently support by this promotion strategy:

- [GitHub](https://github.com/)
- [GitLab](https://gitlab.com/)
- [BitBucket Server / DataCenter](https://www.atlassian.com/software/bitbucket/enterprise)

Select your Git provider via `.spec.promotion.strategy.pull-request.type`. For example, for `gitlab` it would look similar to:

```yaml
promotion:
  strategy:
    pull-request:
      type: gitlab
      url: "https://gitlab.com/weaveworks/<my-awesome-project.git>"
      baseBranch: main
      secretRef:
        name: gitlab-promotion-credentials
```

More info in the [spec](../spec/v1alpha1/pipeline/#pipeline).

### Credentials Secret

In the journey of creating a pull request, there are different secrets involved:

> 1. Pipeline controller receives events via [webhook from leaf clusters](./#setup-notifications-from-leaf-clusters). Hmac is used for authN/authZ so an hmac key should be provided in this case.

Reviewer: ```suggestion
1. Pipeline controller receives events via [webhook from leaf clusters](./#setup-notifications-from-leaf-clusters). An [HMAC](https://en.wikipedia.org/wiki/HMAC) is used for verification so a shared key should be provided in this case.
```
